firmware/main.cpp

```diff
@@ -0,0 +1,140 @@
+#include "mbed.h"
+#include "CAN.h"
+#include "can-ids/CAN_IDS.h"
+
+// Heartbeat message rate in milliseconds
+#define HEARTBEAT_RATE     100ms
+
+#define STEERING_WHEEL_DRS_BYTE 0
+#define BRAKE_PRESSURE_BYTE 0
+#define BRAKE_LIGHT_PRESSURE_CUTOFF 0x80 // TODO Set a reasonable cutoff
+
+// Uncomment to use the LED on the Nucleo dev board instead of the LED pin on the brakelight board.
+#define NUCLEO_LED
+
+#define DEFAULT_BRIGHTNESS 1.0f
+#define MIN_BRIGHTNESS 0.01f
+float brightness = DEFAULT_BRIGHTNESS;
+
+
+Ticker heartbeat;
+CAN can(PA_11, PA_12, 250000);
+char counter = 0;
+
+// Initialise the LEDs as a PWM output
+#ifdef NUCLEO_LED
+PwmOut led(LED1); // Debug led from nucleo board
+#else
+PwmOut led(PA_8); // Transistor for brake light LED
+#endif
+
+// Initialize the servos as outputs
+#define NUM_SERVOS 2 // The configuration for the Nucleo-F445RE doesn't support PWM on pins PA2 and PA3 in MBed.
+// This can likely be fixed if we need more than 2 independent servos.
+PwmOut servos[NUM_SERVOS] = {PwmOut(PA_0), PwmOut(PA_1)};
+
+// Servo positions when DRS is not active (positions in uS, most servos expect 1000-2000 uS range)
+int servo_pos_no_drs[NUM_SERVOS] = {1000, 1000};
+// Servo positions when DRS is active
+int servo_pos_drs[NUM_SERVOS] = {2000, 2000};
+
+
+// Event queue for handling I/O triggered by an ISR (e.g. printing, sending CAN messages)
+EventQueue queue(32 * EVENTS_EVENT_SIZE);
+Thread t;
+
+// Send out a heartbeat message over CAN
+void can_hb() {
+    if(can.write(CANMessage(BRAKE_LIGHT, &counter, 1))) {
+        counter++;
+    } else {
+        printf("Failed to send CAN Heartbeat message :(\n");
```

Comment:
I’m used to mbed 5 where this didn’t work, but I would check that `printf` outputs ok over serial debug. Also, it may be worth adding a `#define PRINT_ERRORS` or something that can be commented out to disable error logging. You can check the old bcm-firmware header for an example of this.

Reply:
I'm using the latest version of mbed and verified that `printf(...)` works over serial for the nucleo board, but it's still a good idea to add a define that removes it since the actual board doesn't have a serial port for debugging.